ENH: Add Cloning

pubpub-zz opened the pull request:
The method `.clone(pdf_dest,[force_duplicate])` clones the objects and all referenced objects.

If an object is already cloned, the already cloned object is returned (unless force_duplicate is set)
mainly for internal use but can be used on a page
for pageObject/DictionnaryObject/[Encoded/Decoded/Content]Stream an extra parameter ignore_fields list that provide the list of fields that should not be cloned.

When available, the pointer to an object is available in `indirect_obj` attribute.

New API for add_page/insert_page that :

* returns the cloned page object
* ignore_fields can be provided as a parameter.

## Others

* file is closed at the end of PdfWriter.write when a filename is provided
* Breaking Change: `add_outline_item` now has a parameter before which is not the last parameter

## Update
* The public API of PdfMerger has been added to PdfWriter (ready to make PdfMerger an alias of it)
* Process properly Outline merging
* Process properly Named destinated

Deals with #1194, #1322, #471, #1337

pubpub-zz commented:
@MartinThoma,
if you can provide some guidelines on how to fix mypy, It will be appreciated.

MartinThoma commented:
I'll have a ok tomorrow :-)

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -296,7 +296,7 @@ def __init__(
         DictionaryObject.__init__(self)
         self.pdf: Optional[PdfReader] = pdf
-        self.indirect_ref = indirect_ref
+        self.indirect_ref = indirect_ref  # type:ignore
```
```suggestion
        self.indirect_ref = indirect_ref  # type: ignore[assignment]
```

pubpub-zz replied:
Thanks, I managed to fix globally mypy yesterday. However your comments are welcomed.
Currently working on adding on "cloning" the "filtered" annotations and articles

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -66,6 +79,36 @@ def hash_value(self) -> bytes:
 
+    def clone(
+        self,
+        pdf_dest: "PdfWriter",  # type: ignore
```
```suggestion
        pdf_dest: "PdfWriter",
```

pubpub-zz replied:
are you sure this is OK for old version?

MartinThoma replied:
Oh, good question. I only checked locally with Python 3.10.2.

Let me upload my changes in another branch to verify (I'll delete that branch later)

MartinThoma replied:
https://github.com/py-pdf/PyPDF2/pull/1396

MartinThoma replied:
Seems to be ok :-)

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -66,6 +79,36 @@ def hash_value(self) -> bytes:
+        """clone object into pdf_dest"""
+        raise Exception("clone PdfObject")
+
+    def _reference_clone(self, clone: Any, pdf_dest: "PdfWriter") -> "PdfObject":  # type: ignore
```
```suggestion
    def _reference_clone(self, clone: Any, pdf_dest: "PdfWriter") -> "PdfObject":
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -81,6 +124,15 @@ def write_to_stream(
 class NullObject(PdfObject):
+    def clone(
+        self,
+        pdf_dest: "PdfWriter",  # type: ignore
```
```suggestion
        pdf_dest: "PdfWriter",
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -112,6 +164,17 @@ class BooleanObject(PdfObject):
 
+    def clone(
+        self,
+        pdf_dest: "PdfWriter",  # type: ignore
```
```suggestion
        pdf_dest: "PdfWriter",
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -160,7 +223,30 @@ def __init__(self, idnum: int, generation: int, pdf: Any) -> None:  # PdfReader
-    def get_object(self) -> Optional[PdfObject]:
+    def clone(
+        self,
+        pdf_dest: "PdfWriter",  # type: ignore
```
```suggestion
        pdf_dest: "PdfWriter",
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -160,7 +223,30 @@ def __init__(self, idnum: int, generation: int, pdf: Any) -> None:  # PdfReader
+        if not force_duplicate and self.idnum in pdf_dest._id_translated[id(self.pdf)]:
+            dup = pdf_dest.get_object(pdf_dest._id_translated[id(self.pdf)][self.idnum])
+        else:
+            dup = self.get_object().clone(pdf_dest, force_duplicate, ignore_fields)  # type: ignore
```
```suggestion
            obj = self.get_object()
            assert obj is not None
            dup = obj.clone(pdf_dest, force_duplicate, ignore_fields)
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -160,7 +223,30 @@ def __init__(self, idnum: int, generation: int, pdf: Any) -> None:  # PdfReader
+        return dup.indirect_ref
+
+    @property
+    def indirect_ref(self) -> "IndirectObject":  # type: ignore
```
```suggestion
    def indirect_ref(self) -> "IndirectObject":  # type: ignore[override]
```

pubpub-zz replied:
fixed

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -128,6 +155,91 @@ def readFromStream(
+                                cur_obj._reference_clone(cur_obj.__class__(), pdf_dest),
+                            )
+                            objs.append((cur_obj, clon))
+                            prev_obj[NameObject(k)] = clon.indirect_ref  # type: ignore
```
```suggestion
                            assert prev_obj is not None
                            prev_obj[NameObject(k)] = clon.indirect_ref
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -128,6 +155,91 @@ def readFromStream(
+            if k not in ignore_fields:
+                if isinstance(v, StreamObject):
+                    if not hasattr(v, "indirect_ref"):
+                        v.indirect_ref = None  # type: ignore
```
```suggestion
                        v.indirect_ref = None  # type: ignore[assignment]
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -128,6 +155,91 @@ def readFromStream(
+                        v.indirect_ref = None  # type: ignore
+                    vv = v.clone(pdf_dest, force_duplicate, ignore_fields)
+                    assert vv.indirect_ref is not None
+                    self[k.clone(pdf_dest)] = vv.indirect_ref  # type: ignore
```
Mypy thinks for some reason that k is of type str, but it is a PdfObject, right?

We could `assert isinstance(k, PdfObject)` which would solve the issue. But this works as well:

```suggestion
                    self[k.clone(pdf_dest)] = vv.indirect_ref  # type: ignore[attr-defined]
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -387,32 +499,61 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
+                inc_parent_counter(self, child_obj.get("/Count", 1))
+                return
+        try:  # insert as first or in the middle
+            prev["/Prev"][NameObject("/Next")] = child  # type: ignore
```
```suggestion
            assert isinstance(prev["/Prev"], DictionaryObject)
            prev["/Prev"][NameObject("/Next")] = child
```

pubpub-zz replied:
is is OK for performance ?

MartinThoma replied:
I don't think it's a big deal. On the one hand, people could remove that by using `python -O` (which I haven't ever seen). On the other hand, the `isinstance` call is cheap:

```
def test_exception():
    try:
        a = 1/0
        return a
    except Exception:
        return 2


if __name__ == '__main__':
    import timeit
    print(f'try-except        : {timeit.timeit("test_exception()", setup="from __main__ import test_exception", number=10**6)*1000:3.0f}ns')
    print(f'isinstance false  : {timeit.timeit("isinstance(foo, int)", setup="""foo="a" """, number=10**6)*1000:3.0f}ns')
    print(f'isinstance true   : {timeit.timeit("isinstance(foo, str)", setup="""foo="a" """, number=10**6)*1000:3.0f}ns')
    print(f'addition int-int  : {timeit.timeit("123456+9876554", number=10**6)*1000:3.0f}ns')
    print(f'addition int-float: {timeit.timeit("123456+9876554.3", number=10**6)*1000:3.0f}ns')
```

gives:

```
try-except        : 287ns
isinstance false  :  98ns
isinstance true   :  73ns
addition int-int  :   9ns
addition int-float:   9ns
```

Our big performance problem is how we deal with floats. The conversion of floats to FloatObjects is dominating pretty much everything.

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -522,7 +664,26 @@ def _reset_node_tree_relationship(child_obj: Any) -> None:
+
+    def _clone(
+        self,
+        src: DictionaryObject,  # type: ignore
```
```suggestion
        src: DictionaryObject,
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -522,7 +664,26 @@ def _reset_node_tree_relationship(child_obj: Any) -> None:
+        """update the object from src"""
+        self._data = cast("StreamObject", src)._data
+        try:
+            if cast("StreamObject", src).decoded_self is None:
+                self.decoded_self = None
+            else:
+                self.decoded_self = cast("StreamObject", src).decoded_self.clone(pdf_dest, True, ignore_fields)  # type: ignore
```
```suggestion
            decoded_self = cast("StreamObject", src).decoded_self
            if decoded_self is None:
                self.decoded_self = None
            else:
                self.decoded_self = decoded_self.clone(pdf_dest, True, ignore_fields)  # type: ignore[assignment]
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -859,8 +1059,9 @@ def read_object(
+        stream.read(-20)
         raise PdfReadError(
-            f"Invalid Elementary Object starting with {tok} @{stream.tell()}"  # type: ignore
+            f"Invalid Elementary Object starting with {tok} @{stream.tell()}: {stream.read(80).__repr__()}"  # type: ignore
```
```suggestion
            f"Invalid Elementary Object starting with {tok!r} @{stream.tell()}: {stream.read(80).__repr__()}"
```

MartinThoma reviewed (commented):
mypy didn't complain when I checked. As you asked me to look at mypy, I checked all 'type: ignore' comments. Several were not necessary at all. In some cases mypy needed an `assert variable is not None` as a hint. And in some cases I could at least narrow the ignore down to be a bit more specific

pubpub-zz commented:
All expected features have been implemented.
Test coverage report is not so good : still not understanding codecov  😑
I would not propose to make PdfMerger an alias of PdfWriter yet, but just indicate that PdfWriter can be used instead
@MartinThoma give me your opinion on the PR

MartinThoma commented:
> I would **not** propose to make PdfMerger an alias of PdfWriter yet, but just indicate that PdfWriter can be used instead

I agree on that one. However, we can change examples in the documentation to use PdfWriter everywhere and add documentation for common examples how the PdfWriter can replace PdfMerger (and why people should do that)

MartinThoma commented on tests/test_generic.py:
```diff
@@ -472,23 +474,25 @@ def test_TextStringObject_autodetect_utf16():
 
 
 def test_remove_child_not_in_tree():
+    class ChildDummy(DictionaryObject):
+        @property
+        def indirect_ref(self):
+            return self
+
```
You could move this class outside of the test. This way you would not have to repeat it in the other tests.

Maybe adding a docstring to the class would then be a good idea:

```
    class ChildDummy(DictionaryObject):
         """A helper class for testing interactions with the TreeObject."""
        @property
        def indirect_ref(self):
            return self
```

pubpub-zz replied:
isolated into #1411

MartinThoma commented on tests/test_merger.py:
```diff
@@ -117,6 +119,7 @@ def check_outline(tmp_path):
         "Bar",
         "Baz",
         "foo",
+        "An outline item",  # this has been moved to end normal???
```
I think this is something where I need to have a closer look. This could indicate an issue.

pubpub-zz replied:
isolated into #1411

MartinThoma commented on tests/test_writer.py:
```diff
@@ -478,15 +479,18 @@ def test_add_named_destination():
 
-    writer.add_named_destination(NameObject("A named dest"), 2)
-    writer.add_named_destination(NameObject("A named dest2"), 2)
+    writer.add_named_destination(TextStringObject("A named dest"), 2)
+    writer.add_named_destination(TextStringObject("A named dest2"), 2)
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
isolated into #1411

MartinThoma commented on tests/test_merger.py:
```diff
@@ -96,8 +99,8 @@ def merger_operate(merger):
 
-    merger.add_metadata({"author": "Martin Thoma"})
-    merger.add_named_destination("title", 0)
+    merger.add_metadata({"/Author": "Martin Thoma"})
+    merger.add_named_destination("/Title", 0)
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
isolated into #1411

MartinThoma commented on PyPDF2/pagerange.py:
```diff
@@ -170,4 +170,4 @@ def parse_filename_page_ranges(
 
 
-PageRangeSpec = Union[str, PageRange, Tuple[int, int], Tuple[int, int, int]]
+PageRangeSpec = Union[str, PageRange, Tuple[int, int], Tuple[int, int, int], List[int]]
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
code was quite often too tight. For this it's OK will do

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -296,7 +298,7 @@ def __init__(
         DictionaryObject.__init__(self)
         self.pdf: Optional[PdfReader] = pdf
-        self.indirect_ref = indirect_ref
+        self.indirect_ref = indirect_ref  # type: ignore[assignment]
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
code reviewed:  a PdfObject indirect_ref can be set to None, to force referencing the object during cloning.

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -859,8 +1059,9 @@ def read_object(
             return NumberObject.read_from_stream(stream)
     else:
+        stream.read(-20)
         raise PdfReadError(
-            f"Invalid Elementary Object starting with {tok} @{stream.tell()}"  # type: ignore
+            f"Invalid Elementary Object starting with {tok!r} @{stream.tell()}: {stream.read(80).__repr__()}"
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
transfered in #1412

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -1041,7 +1247,7 @@ def __init__(
     ) -> None:
         DictionaryObject.__init__(self)
-        self[NameObject("/Title")] = title
+        self[NameObject("/Title")] = TextStringObject(title)
```
This is a change that is not related to cloning, right? It was just wrong in the tests before.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
included in #1411

MartinThoma commented on PyPDF2/_merger.py:
```diff
@@ -649,7 +649,7 @@ def add_outline_item(
             raise RuntimeError(ERR_CLOSED_WRITER)
         return writer.add_outline_item(
-            title, pagenum, parent, color, bold, italic, fit, *args
+            title, pagenum, parent, None, color, bold, italic, fit, *args
```
Why is this necessary?

pubpub-zz replied:
add_outline now proposes Before parameter (to be set to None if you wan to append at the end of the list). Although the parameter is fully optional, I prefer to be clear

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -128,8 +134,9 @@ class (typically :class:`PdfReader<PyPDF2.PdfReader>`).
     def __init__(self, fileobj: StrByteType = "") -> None:
         self._header = b"%PDF-1.3"
-        self._objects: List[Optional[PdfObject]] = []  # array of indirect objects
+        self._objects: List[PdfObject] = []  # array of indirect objects
```
This is a change that is not related to cloning, right? It seems to be an unrelated improvement you made while writing this PR.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

pubpub-zz replied:
transfered into #1411

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -196,16 +202,23 @@ def pdf_header(self) -> bytes:
         return self._objects[ido.idnum - 1]  # type: ignore
 
-    def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
+    def getObject(
+        self, ido: Union[int, IndirectObject]
+    ) -> PdfObject:  # pragma: no cover
```
This is a change that is not related to cloning, right? It seems to be an unrelated improvement you made while writing this PR.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)

MartinThoma commented:
@pubpub-zz I see quite a lot of improvements which are not directly related to the core feature you've implemented.

In order to make this PR easier to review, would it be OK for you if I make a second PR that already merges those smaller improvements?

To give you attribution for those changes, I would use the co-authored-by feature of GitHub + write in the commit that this is solely your work

pubpub-zz commented:
@Trigve,
discussion in the PR looks good 😊

Trigve commented:
The one thing that looks suspicious to me are recursive imports (`generic/_base.py` -> `_writer.py` -> `generic/__init__.py` -> `generic/base.py`) of PdfWriter (I know it is only annotation issue but every recursive import smells of some code design issue). Why not creating new "interface" class (something like `IPdfWriteable`) and using it there?  `PdfObject._reference_clone()` could be simplified then (no need to use "internal" stuff of `PdfWriter`) I think.

pubpub-zz commented:
> The one thing that looks suspicious to me are recursive imports (`generic/_base.py` -> `_writer.py` -> `generic/__init__.py` -> `generic/base.py`) of PdfWriter (I know it is only annotation issue but every recursive import smells of some code design issue). Why not creating new "interface" class (something like `IPdfWriteable`) and using it there? `PdfObject._reference_clone()` could be simplified then (no need to use "internal" stuff of `PdfWriter`) I think.

I agree with you,
Actually the main reason about this recursive call is that I'm not so good at annotation 😚
Your comment give me some good advice I have some refactoring to extract some side code. I will try to implement it as proposed.

MartinThoma commented:
@pubpub-zz What do you think about my suggestion to move the improvements that are not related to cloning out? I can do that if you want.

pubpub-zz commented:
In progress... Hope to have something on Saturday

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -94,7 +110,9 @@ def clone(
 
-    def _reference_clone(self, clone: Any, pdf_dest: "PdfWriter") -> "PdfObject":
+    def _reference_clone(
+        self, clone: Any, pdf_dest: _PdfWriterInterface
```
Here the interface is fine as it's only PyPDF2 internal

pubpub-zz replied:
reviewed to ease readability

manathan1984 commented:
Looking forward to the feature being added. Also, please point to the updated documentation on how to use it properly when completed.

pubpub-zz commented on PyPDF2/generic/_base.py:
```diff
@@ -81,7 +97,7 @@ def hash_value(self) -> bytes:
     def clone(
         self,
-        pdf_dest: "PdfWriter",
+        pdf_dest: _PdfWriterInterface,
```
I understand your position. I've reviewed the code to create a local aliasing to ease reading

MartinThoma commented:
@pubpub-zz I've merged the two smaller PRs. Would you mind to resolve the conflicts / update the PR?

MartinThoma commented:
Hi @pubpub-zz,

The PR again has several merge conflicts and is pretty huge. Just to give you some feedback:

Here I would like to hear others opinions (e.g. @MasterOdin if you have time?):

* _PdfWriterInterface, _PdfDocumentInterface: I think using Protocols (https://peps.python.org/pep-0544/) might be cleaner. They are only available in Python 3.8+ though :-/
* test_merger.py::check_outline : I think there might be a bug?

Then there are a lot of changes that I would directly merge (e.g. if you want to make another PR with them / would allow me to do it). That would prevent merge conflicts and make reviewing easier. The changes are:

* reader.threads
* Various smaller changes:
    * Removal of newline in PyPDF2/generic/_data_structures.py
    * pagerange.py
    * test_generic.py: The ChildDummy stuff
    * test_merger.py: `merger_operate`
    * test_reader.py
    * test_writer.py

pubpub-zz commented:
@MartinThoma 
I've resolved the conflicts, the documentation has been updated. 
still some work should be done to improve the test ( I have some trouble to find how to improve that) 
Can you check that what you are expecting to be fixed is ok for you

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1204,8 +1279,9 @@ def addBookmarkDict(
-        pagenum: int,
+        pagenum: Union[None, PageObject, IndirectObject, int],
         parent: Union[None, TreeObject, IndirectObject] = None,
+        before: Union[None, TreeObject, IndirectObject] = None,
```
This is a breaking change as it changes the order of parameters. Would it be possible to move the new parameter to the end?

pubpub-zz replied:
not really : there is a check on the type of italic that allows to detect weither "before" parameter is present or not: the new function is compatible with previous arguments.

pubpub-zz replied:
@MartinThoma OK for you ?

MasterOdin commented:
> * _PdfWriterInterface, _PdfDocumentInterface: I think using Protocols ([peps.python.org/pep-0544](https://peps.python.org/pep-0544/)) might be cleaner. They are only available in Python 3.8+ though :-/

My understanding was that protocols are available in Python 3.4-3.7 from the `typings_extension` package. It's not clear to me why you added them in #1303 and then was removed in #1309 with a vague statement about pytest locally failing, though it wasn't failing on CI...

I would advocate for using protocols here, and just name it `PdfWriterProtocol` and use it as such and remove the weird `PdfWriter` alias. Granted, the larger change would be to retype everywhere to use the `PdfWriterProtocol` and not have a weird mix of `PdfWriter` and `PdfWriterProtocol`, with a comment above explaining why it exists, and a link to the base `PdfWriter` class.

> Then there are a lot of changes that I would directly merge (e.g. if you want to make another PR with them / would allow me to do it). That would prevent merge conflicts and make reviewing easier. The changes are:
> 
> * reader.threads
> * Various smaller changes:
>   
>   * Removal of newline in PyPDF2/generic/_data_structures.py
>   * pagerange.py
>   * test_generic.py: The ChildDummy stuff
>   * test_merger.py: `merger_operate`
>   * test_reader.py
>   * test_writer.py

I will always advocate that PRs should be split apart when possible, even if it means that you might end up in the scenario where a final PR (e.g. this one) relies on some of the broken off smaller pieces. My view is that if you ever have to revert a PR for whatever reason, the smaller it is, the less work you'd lose and also have to re-deal with in a new fixed PR.

pubpub-zz commented:
the removal in #1309 was not intended mostlikely due to invalid rebasing. I've re-introduced it.
I had to create a new file for protocol to prevent circular import.

pubpub-zz commented:
@MartinThoma / @MasterOdin 
Can you tell me if the use of purpose is OK?

MartinThoma commented:
> Can you tell me if the use of purpose is OK?

@pubpub-zz I guess "purpose" was auto-corrected and should be "protocols"?

Yes, you can use protocols:

```python
try:
    from typing import Protocol
except ImportError:
    # PEP-544: Python 3.6 and 3.7
    from typing_extensions import Protocol
```

xilopaint commented:
Hey guys, is this PR close to being merged?

pubpub-zz commented on PyPDF2/_writer.py:
```diff
@@ -215,20 +229,41 @@ def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
-        if page.pdf is not None:
-            other = page.pdf.pdf_header
+        page_org = page
+        if excluded_keys is None:
```
Agree

MartinThoma commented:
Sorry, this week is crazy for me: Moving apartments + my wedding. Next week I'm on a business trip. I hope we can merge it this year :-)

pubpub-zz commented:
> Then there are a lot of changes that I would directly merge (e.g. if you want to make another PR with them / would allow me to do it). That would prevent merge conflicts and make reviewing easier. The changes are:
> 
>     * reader.threads

Extracted in #1458 although quite meaningless for the moment
> 
>     * Various smaller changes:
>       
>       * Removal of newline in PyPDF2/generic/_data_structures.py

Removed

>       * pagerange.py

PageRangeSpec extracted in #1456

>       * test_generic.py: The ChildDummy stuff

extracted  in #1411

>       * test_merger.py: `merger_operate`

?? not understood

>       * test_reader.py

moved to PR#1459

>       * test_writer.py

move some into PR#1459 but very limited

pubpub-zz commented:
@MartinThoma 
all yours

MartinThoma commented:
I'll merge this PR this weekend. I need to check what I did wrong with the CI though

MartinThoma commented on PyPDF2/_reader.py:
```diff
@@ -1391,6 +1392,8 @@ def cache_indirect_object(
             logger_warning(msg, __name__)
         self.resolved_objects[(generation, idnum)] = obj
+        if obj is not None:
+            obj.indirect_ref = IndirectObject(idnum, generation, self)
```
```suggestion
            obj.indirect_reference = IndirectObject(idnum, generation, self)
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -198,8 +204,11 @@ def pdf_header(self, new_header: bytes) -> None:
         self._header = new_header
 
     def _add_object(self, obj: PdfObject) -> IndirectObject:
+        if hasattr(obj, "indirect_ref") and obj.indirect_ref.pdf == self:  # type: ignore
+            return obj.indirect_ref  # type: ignore
```
```suggestion
        if hasattr(obj, "indirect_reference") and obj.indirect_reference.pdf == self:  # type: ignore
            return obj.indirect_reference  # type: ignore
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -198,8 +204,11 @@ def pdf_header(self, new_header: bytes) -> None:
+            return obj.indirect_ref  # type: ignore
         self._objects.append(obj)
-        return IndirectObject(len(self._objects), 0, self)
+        obj.indirect_ref = IndirectObject(len(self._objects), 0, self)
+        return obj.indirect_ref
```
```suggestion
        obj.indirect_reference = IndirectObject(len(self._objects), 0, self)
        return obj.indirect_reference
```

MartinThoma commented on PyPDF2/_protocols.py:
```diff
@@ -0,0 +1,65 @@
+
+
+class PdfObjectProtocol(Protocol):
+    indirect_ref: Any
```
```suggestion
    indirect_reference: Any
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -50,9 +51,10 @@
+class PdfObject(PdfObjectProtocol):
     # function for calculating a hash value
     hash_func: Callable[..., "hashlib._Hash"] = hashlib.sha1
+    indirect_ref: Optional["IndirectObject"]
```
```suggestion
    indirect_reference: Optional["IndirectObject"]
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -66,6 +68,50 @@ def hash_value(self) -> bytes:
+        self, clone: Any, pdf_dest: PdfWriterProtocol
+    ) -> PdfObjectProtocol:
+        """
+        reference the object within the _objects of pdf_dest only if indirect_ref attribute exists (which means the objects was already identified in xref/xobjstm)
+        if object has been already referenced do nothing
+        """
+        try:
+            if clone.indirect_ref.pdf == pdf_dest:
+                return clone
+        except Exception:
+            pass
+        if hasattr(self, "indirect_ref"):
+            ind = self.indirect_ref
+            i = len(pdf_dest._objects) + 1
+            if ind is not None:
+                if id(ind.pdf) not in pdf_dest._id_translated:
+                    pdf_dest._id_translated[id(ind.pdf)] = {}
+                if ind.idnum in pdf_dest._id_translated[id(ind.pdf)]:
+                    obj = pdf_dest.get_object(
+                        pdf_dest._id_translated[id(ind.pdf)][ind.idnum]
+                    )
+                    assert obj is not None
+                    return obj
+                pdf_dest._id_translated[id(ind.pdf)][ind.idnum] = i
+            pdf_dest._objects.append(clone)
+            clone.indirect_ref = IndirectObject(i, 0, pdf_dest)
+        return clone
```
```suggestion
        reference the object within the _objects of pdf_dest only if
        indirect_reference attribute exists (which means the objects
        was already identified in xref/xobjstm)
        if object has been already referenced do nothing
        """
        try:
            if clone.indirect_reference.pdf == pdf_dest:
                return clone
        except Exception:
            pass
        if hasattr(self, "indirect_reference"):
            ind = self.indirect_reference
            i = len(pdf_dest._objects) + 1
            if ind is not None:
                if id(ind.pdf) not in pdf_dest._id_translated:
                    pdf_dest._id_translated[id(ind.pdf)] = {}
                if ind.idnum in pdf_dest._id_translated[id(ind.pdf)]:
                    obj = pdf_dest.get_object(
                        pdf_dest._id_translated[id(ind.pdf)][ind.idnum]
                    )
                    assert obj is not None
                    return obj
                pdf_dest._id_translated[id(ind.pdf)][ind.idnum] = i
            pdf_dest._objects.append(clone)
            clone.indirect_reference = IndirectObject(i, 0, pdf_dest)
        return clone
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -636,13 +686,10 @@ def append_pages_from_reader(
-            reader_page = reader.pages[reader_page_number]
-            self.add_page(reader_page)
-            writer_page = self.pages[writer_num_pages + reader_page_number]
+            reader_page = reader.pages[rpagenum]
```
```suggestion
            reader_page = reader.pages[reader_page_number]
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1441,14 +1512,31 @@ def addBookmark(
             "This method is not yet implemented. Use :meth:`add_outline_item` instead."
         )
 
+
```
```suggestion
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1441,14 +1512,31 @@ def addBookmark(
+        nd.extend([TextStringObject(title), destination])
+        return
+
+
```
```suggestion
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1470,14 +1558,21 @@ def add_named_destination_object(
+    def add_named_destination_object(self, page_destination: Destination) -> IndirectObject:
+        page_destination_ref = self._add_object(page_destination.dest_array)
+        self.add_named_destination_array(
+            cast("TextStringObject", dest["/Title"]), page_destination_ref
```
```suggestion
            cast("TextStringObject", page_destination["/Title"]), page_destination_ref
```

MartinThoma commented:
@pubpub-zz I've messed your PR up :see_no_evil:  The first `add_named_destination_object` should be `add_named_destination_array`, but GitHub doesn't allow me to change it. Could you please adjust that?

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1356,48 +1412,60 @@ def add_outline_item(
-        :param tuple color: Color of the outline item's font as a red, green, blue tuple
+        :param parent: A reference to a parent outline item to create nested
+            outline items.
+        :param tuple colo r: Color of the outline item's font as a red, green, blue tuple
```
```suggestion
        :param tuple color: Color of the outline item's font as a red, green, blue tuple
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1420,6 +1488,7 @@ def add_bookmark(
             title,
             pagenum,
             parent,
+            None, c
```
```suggestion
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1458,7 +1527,22 @@ def add_outline(self) -> None:
+        nd.extend([TextStringObject(title), destination])
+        return
+
+    def add_named_destination_array(
```
```suggestion
    def add_named_destination_object(
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1479,14 +1563,21 @@ def add_named_destination_object(
 
-        nd = self.get_named_dest_root()
-        nd.extend([page_destination["/Title"], page_destination_ref])  # type: ignore
+    def add_named_destination_object(self, page_destination: Destination) -> IndirectObject:
+        page_destination_ref = self._add_object(page_destination.dest_array)
+        self.add_named_destination_array(
+            cast("TextStringObject", page_destination["/Title"]), page_destination_ref
+        )
```
```suggestion
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1479,14 +1563,21 @@ def add_named_destination_object(
+        self.add_named_destination_array(
+            cast("TextStringObject", dest["/Title"]), page_destination_ref
+        )
+        return page_destination_ref
```
```suggestion
```

MartinThoma commented on tests/test_writer.py:
```diff
@@ -512,11 +512,18 @@ def test_add_named_destination():
     assert root[1].pdf == writer
     assert root[1].get_object()["/S"] == NameObject("/GoTo")
-    assert root[1].get_object()["/D"][0].get_object() == writer.pages[2]
+    assert root[1].get_object()["/D"][0] == writer.pages[2].indirect_ref
```
```suggestion
    assert root[1].get_object()["/D"][0] == writer.pages[2].indirect_reference
```

MartinThoma commented on tests/test_writer.py:
```diff
@@ -512,11 +512,18 @@ def test_add_named_destination():
     assert root[3].pdf == writer
     assert root[3].get_object()["/S"] == NameObject("/GoTo")
-    assert root[3].get_object()["/D"][0].get_object() == writer.pages[2]
+    assert root[3].get_object()["/D"][0] == writer.pages[2].indirect_ref
```
```suggestion
    assert root[3].get_object()["/D"][0] == writer.pages[2].indirect_reference
```

MartinThoma commented on tests/test_writer.py:
```diff
@@ -512,11 +512,18 @@ def test_add_named_destination():
+
+    assert writer.get_object(root[1].idnum) == writer.get_object(root[1])
+    with pytest.raises(ValueError) as exc:
+        writer.get_object(reader.pages[0].indirect_ref)
```
```suggestion
        writer.get_object(reader.pages[0].indirect_reference)
```

MartinThoma commented on tests/test_writer.py:
```diff
@@ -882,10 +882,78 @@ def test_startup_dest():
+    writer.append(reader, (0, 10))
+    assert len(writer._objects) >= nb + 200
+    nb = len(writer._objects)
+    writer.reset_translation(reader.pages[0].indirect_ref)
```
```suggestion
    writer.reset_translation(reader.pages[0].indirect_reference)
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -235,20 +248,38 @@ def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
         pages = cast(DictionaryObject, self.get_object(self._pages))
-        action(pages[PA.KIDS], page_ind)
+        assert page.indirect_ref is not None
+        action(pages[PA.KIDS], page.indirect_ref)
```
```suggestion
        action(pages[PA.KIDS], page.indirect_reference)
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+        nthread = thread.clone(
+            self, force_duplicate=True, ignore_fields=("/F",)
+        )  # use of clone to keep link between reader and writer
+        self.threads.append(nthread.indirect_ref)
```
```suggestion
        self.threads.append(nthread.indirect_reference)
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                pag_obj = cast("PageObject", pag.get_object())
+                if "/B" not in pag_obj:
+                    pag_obj[NameObject("/B")] = ArrayObject()
+                cast("ArrayObject", pag_obj["/B"]).append(new_article.indirect_ref)
```
```suggestion
                cast("ArrayObject", pag_obj["/B"]).append(new_article.indirect_reference)
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -74,6 +75,33 @@
+    ) -> "ArrayObject":
+        """clone object into pdf_dest"""
+        try:
+            if self.indirect_ref.pdf == pdf_dest and not force_duplicate:  # type: ignore
```
```suggestion
            if self.indirect_reference.pdf == pdf_dest and not force_duplicate:  # type: ignore
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+    ) -> "DictionaryObject":
+        """clone object into pdf_dest"""
+        try:
+            if self.indirect_ref.pdf == pdf_dest and not force_duplicate:  # type: ignore
```
```suggestion
            if self.indirect_reference.pdf == pdf_dest and not force_duplicate:  # type: ignore
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -693,21 +858,58 @@ def __init__(
+    ) -> "ContentStream":
+        """clone object into pdf_dest"""
+        try:
+            if self.indirect_ref.pdf == pdf_dest and not force_duplicate:  # type: ignore
```
```suggestion
            if self.indirect_reference.pdf == pdf_dest and not force_duplicate:  # type: ignore
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    obj3 = obj2.clone(writer)
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_ref == obj3.indirect_ref
+    obj3 = obj2.indirect_ref.clone(writer)
```
```suggestion
    obj3 = obj2.indirect_reference.clone(writer)
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_ref == obj3.indirect_ref
+
+    obj3 = obj2.indirect_ref.clone(writer, True)
```
```suggestion
    obj3 = obj2.indirect_reference.clone(writer, True)
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -74,6 +75,33 @@
+                dup = data._reference_clone(
+                    data.clone(pdf_dest, force_duplicate, ignore_fields), pdf_dest
+                )
+                arr.append(dup.indirect_ref)
```
```suggestion
                arr.append(dup.indirect_reference)
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -160,7 +228,34 @@ def __init__(self, idnum: int, generation: int, pdf: Any) -> None:  # PdfReader
+            assert obj is not None
+            dup = obj.clone(pdf_dest, force_duplicate, ignore_fields)
+        assert dup is not None
+        assert dup.indirect_ref is not None
+        return dup.indirect_ref
```
```suggestion
        assert dup.indirect_reference is not None
        return dup.indirect_reference
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -74,6 +75,33 @@
+        arr = cast("ArrayObject", self._reference_clone(ArrayObject(), pdf_dest))
+        for data in self:
+            if isinstance(data, StreamObject):
+                # if not hasattr(data, "indirect_ref"):
```
```suggestion
                # if not hasattr(data, "indirect_reference"):
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+            if k not in ignore_fields:
+                if isinstance(v, StreamObject):
+                    if not hasattr(v, "indirect_ref"):
+                        v.indirect_ref = None
```
```suggestion
                        v.indirect_reference = None
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+                    if not hasattr(v, "indirect_ref"):
+                        v.indirect_ref = None
+                    vv = v.clone(pdf_dest, force_duplicate, ignore_fields)
+                    assert vv.indirect_ref is not None
```
```suggestion
                    assert vv.indirect_reference is not None
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+                        v.indirect_ref = None
+                    vv = v.clone(pdf_dest, force_duplicate, ignore_fields)
+                    assert vv.indirect_ref is not None
+                    self[k.clone(pdf_dest)] = vv.indirect_ref  # type: ignore[attr-defined]
```
```suggestion
                    self[k.clone(pdf_dest)] = vv.indirect_reference  # type: ignore[attr-defined]
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -388,33 +502,63 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
         child_obj = child.get_object()
-        child = pdf.get_reference(child_obj)
-        assert isinstance(child, IndirectObject)
+        child = child.indirect_ref  # get_reference(child_obj)
```
```suggestion
        child = child.indirect_reference  # get_reference(child_obj)
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -388,33 +502,63 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
-        child_obj[NameObject("/Parent")] = parent_ref
+            prev = cast("DictionaryObject", self["/Last"])
+
+        while prev.indirect_ref != before:
```
```suggestion
        while prev.indirect_reference != before:
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -473,8 +475,8 @@ def test_remove_child_not_in_that_tree():
     tree.indirect_reference = NullObject()
-    # child = ChildDummy(TreeObject())
     child = TreeObject()
+    child.indirect_ref = NullObject()
```
```suggestion
    child.indirect_reference = NullObject()
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    assert "clone PdfObject" in exc.value.args[0]
+
+    obj1 = DictionaryObject()
+    obj1.indirect_ref = None
```
```suggestion
    obj1.indirect_reference = None
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    assert len(writer._objects) == n + 1
+    obj3 = obj2.clone(writer)
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_ref == obj3.indirect_ref
+    obj3 = obj2.indirect_reference.clone(writer)
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_ref == obj3.indirect_ref
+    assert obj2.indirect_ref == obj2._reference_clone(obj2, writer).indirect_ref
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_ref == obj3.indirect_ref
+
+    obj3 = obj2.indirect_reference.clone(writer, True)
+    assert len(writer._objects) == n + 2
+    assert obj2.indirect_ref != obj3.indirect_ref
```
```suggestion
    assert obj2.indirect_reference == obj3.indirect_ref
    obj3 = obj2.indirect_reference.clone(writer)
    assert len(writer._objects) == n + 1
    assert obj2.indirect_reference == obj3.indirect_ref
    assert obj2.indirect_reference == obj2._reference_clone(obj2, writer).indirect_reference
    assert len(writer._objects) == n + 1
    assert obj2.indirect_reference == obj3.indirect_reference

    obj3 = obj2.indirect_reference.clone(writer, True)
    assert len(writer._objects) == n + 2
    assert obj2.indirect_reference != obj3.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+        srcpages = {}
+        for i in pages:
+            pg = reader.pages[i]
+            assert pg.indirect_ref is not None
```
```suggestion
            assert pg.indirect_reference is not None
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                cast("ArrayObject", pag_obj["/B"]).append(new_article.indirect_ref)
+            current_article = cast("DictionaryObject", current_article["/N"])
+            if current_article == first_article:
+                new_article[NameObject("/N")] = new_first.indirect_ref  # type: ignore
+                new_first[NameObject("/V")] = new_article.indirect_ref  # type: ignore
+                current_article = None
+        assert nthread.indirect_ref is not None
```
```suggestion
                new_article[NameObject("/N")] = new_first.indirect_reference  # type: ignore
                new_first[NameObject("/V")] = new_article.indirect_reference  # type: ignore
                current_article = None
        assert nthread.indirect_reference is not None
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+        elif isinstance(page, IndirectObject):
+            _i = page
+        try:
+            return pages[_i.idnum].indirect_ref  # type: ignore
```
```suggestion
            return pages[_i.idnum].indirect_reference  # type: ignore
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+
+        i = 0
+        while o is not None:
+            if o.indirect_ref == outline_item or o.get("/Title", None) == outline_item:
```
```suggestion
            if o.indirect_reference == outline_item or o.get("/Title", None) == outline_item:
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -235,20 +248,38 @@ def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
-        page_ind = self._add_object(page)
         pages = cast(DictionaryObject, self.get_object(self._pages))
-        action(pages[PA.KIDS], page_ind)
+        assert page.indirect_ref is not None
```
```suggestion
        assert page.indirect_reference is not None
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+        for k, v in src.items():
+            if k not in ignore_fields:
+                if isinstance(v, StreamObject):
+                    if not hasattr(v, "indirect_ref"):
```
```suggestion
                    if not hasattr(v, "indirect_reference"):
```

MartinThoma commented on PyPDF2/generic/_base.py:
```diff
@@ -160,7 +228,34 @@ def __init__(self, idnum: int, generation: int, pdf: Any) -> None:  # PdfReader
+        return dup.indirect_reference
+
+    @property
+    def indirect_ref(self) -> "IndirectObject":  # type: ignore[override]
```
```suggestion
    def indirect_reference(self) -> "IndirectObject":  # type: ignore[override]
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -221,11 +230,15 @@ def get_object(
         assert (
             indirect_reference is not None
         )  # the None value is only there to keep the deprecated name
+        if isinstance(ido, int):
```
```suggestion
        if isinstance(indirect_reference, int):
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -221,11 +230,15 @@ def get_object(
             indirect_reference is not None
         )  # the None value is only there to keep the deprecated name
+        if isinstance(ido, int):
+            return self._objects[ido - 1]
```
```suggestion
            return self._objects[indirect_reference - 1]
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -457,6 +601,7 @@ def _remove_node_from_tree(
 
     def remove_child(self, child: Any) -> None:
         child_obj = child.get_object()
+        child = child_obj.indirect_ref
```
```suggestion
        child = child_obj.indirect_reference
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -388,33 +502,63 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
             self[NameObject("/Count")] = NumberObject(0)
-            prev = None
+            self[NameObject("/Last")] = child
+            child_obj[NameObject("/Parent")] = self.indirect_ref
```
```suggestion
            child_obj[NameObject("/Parent")] = self.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -235,20 +248,38 @@ def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
+        # therefore in order to add easily multiple copies of the same page, we need to create a new
+        # dictionary for the page, however the objects below (including content) is not duplicated
+        try:  # delete an already existing page
+            del self._id_translated[id(page_org.indirect_ref.pdf)][  # type: ignore
+                page_org.indirect_ref.idnum  # type: ignore
+            ]
```
```suggestion
            del self._id_translated[id(page_org.indirect_reference.pdf)][  # type: ignore
                page_org.indirect_reference.idnum  # type: ignore
            ]
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+            pg = reader.pages[i]
+            assert pg.indirect_reference is not None
+            if position is None:
+                srcpages[pg.indirect_ref.idnum] = self.add_page(
+                    pg, list(excluded_fields) + ["/B", "/Annots"]  # type: ignore
+                )
+            else:
+                srcpages[pg.indirect_ref.idnum] = self.insert_page(
+                    pg, position, list(excluded_fields) + ["/B", "/Annots"]  # type: ignore
+                )
+                position += 1
+            srcpages[pg.indirect_ref.idnum].original_page = pg
```
```suggestion
                srcpages[pg.indirect_reference.idnum] = self.add_page(
                    pg, list(excluded_fields) + ["/B", "/Annots"]  # type: ignore
                )
            else:
                srcpages[pg.indirect_reference.idnum] = self.insert_page(
                    pg, position, list(excluded_fields) + ["/B", "/Annots"]  # type: ignore
                )
                position += 1
            srcpages[pg.indirect_reference.idnum].original_page = pg
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+            # try:
+            if isinstance(dest["/Page"], NullObject):
+                pass  # self.add_named_destination_array(dest["/Title"],arr)
+            elif dest["/Page"].indirect_ref.idnum in srcpages:
+                arr[NumberObject(0)] = srcpages[
+                    dest["/Page"].indirect_ref.idnum
```
```suggestion
            elif dest["/Page"].indirect_reference.idnum in srcpages:
                arr[NumberObject(0)] = srcpages[
                    dest["/Page"].indirect_reference.idnum
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1192,15 +1245,11 @@ def get_named_dest_root(self) -> ArrayObject:
-            idnum = self._objects.index(names) + 1
-            names_ref = IndirectObject(idnum, 0, self)
-            assert names_ref.get_object() == names
+            names_ref = names.indirect_ref
```
```suggestion
            names_ref = names.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1192,15 +1245,11 @@ def get_named_dest_root(self) -> ArrayObject:
-                idnum = self._objects.index(dests) + 1
-                dests_ref = IndirectObject(idnum, 0, self)
-                assert dests_ref.get_object() == dests
+                dests_ref = dests.indirect_ref
```
```suggestion
                dests_ref = dests.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                        self._add_object(DictionaryObject()).get_object(),
+                    )
+                    new_first = new_article
+                    nthread[NameObject("/F")] = new_article.indirect_ref
```
```suggestion
                    nthread[NameObject("/F")] = new_article.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                        "DictionaryObject",
+                        self._add_object(
+                            DictionaryObject(
+                                {NameObject("/V"): new_article.indirect_ref}
```
```suggestion
                                {NameObject("/V"): new_article.indirect_reference}
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                            )
+                        ).get_object(),
+                    )
+                    new_article[NameObject("/N")] = new_article2.indirect_ref
+                    new_article = new_article2
+                new_article[NameObject("/P")] = pag
+                new_article[NameObject("/T")] = nthread.indirect_ref
+                new_article[NameObject("/R")] = current_article["/R"]
+                pag_obj = cast("PageObject", pag.get_object())
+                if "/B" not in pag_obj:
+                    pag_obj[NameObject("/B")] = ArrayObject()
+                cast("ArrayObject", pag_obj["/B"]).append(new_article.indirect_reference)
+            current_article = cast("DictionaryObject", current_article["/N"])
+            if current_article == first_article:
+                new_article[NameObject("/N")] = new_first.indirect_reference  # type: ignore
+                new_first[NameObject("/V")] = new_article.indirect_reference  # type: ignore
+                current_article = None
+        assert nthread.indirect_reference is not None
+        return nthread.indirect_ref
```
```suggestion
                    new_article[NameObject("/N")] = new_article2.indirect_reference
                    new_article = new_article2
                new_article[NameObject("/P")] = pag
                new_article[NameObject("/T")] = nthread.indirect_reference
                new_article[NameObject("/R")] = current_article["/R"]
                pag_obj = cast("PageObject", pag.get_object())
                if "/B" not in pag_obj:
                    pag_obj[NameObject("/B")] = ArrayObject()
                cast("ArrayObject", pag_obj["/B"]).append(new_article.indirect_reference)
            current_article = cast("DictionaryObject", current_article["/N"])
            if current_article == first_article:
                new_article[NameObject("/N")] = new_first.indirect_reference  # type: ignore
                new_first[NameObject("/V")] = new_article.indirect_reference  # type: ignore
                current_article = None
        assert nthread.indirect_reference is not None
        return nthread.indirect_reference
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    assert obj2.indirect_reference == obj3.indirect_ref
+    obj3 = obj2.indirect_reference.clone(writer)
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_reference == obj3.indirect_ref
```
```suggestion
    assert obj2.indirect_reference == obj3.indirect_reference
```

MartinThoma commented on tests/test_generic.py:
```diff
@@ -943,3 +945,48 @@ def test_create_string_object_force():
+    assert len(writer._objects) == n + 1
+    obj3 = obj2.clone(writer)
+    assert len(writer._objects) == n + 1
+    assert obj2.indirect_reference == obj3.indirect_ref
```
```suggestion
    assert obj2.indirect_reference == obj3.indirect_reference
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -388,33 +502,63 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
+            del child_obj["/Next"]
+        child_obj[NameObject("/Next")] = prev
+        prev[NameObject("/Prev")] = child
+        child_obj[NameObject("/Parent")] = self.indirect_ref
```
```suggestion
        child_obj[NameObject("/Parent")] = self.indirect_reference
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -388,33 +502,63 @@ def addChild(self, child: Any, pdf: Any) -> None:  # pragma: no cover
+                prev = cast("TreeObject", prev["/Next"])
+            else:  # append at the end
+                prev[NameObject("/Next")] = cast("TreeObject", child)
+                child_obj[NameObject("/Prev")] = prev.indirect_ref
+                child_obj[NameObject("/Parent")] = self.indirect_ref
```
```suggestion
                child_obj[NameObject("/Prev")] = prev.indirect_reference
                child_obj[NameObject("/Parent")] = self.indirect_reference
```

MartinThoma commented on PyPDF2/generic/_data_structures.py:
```diff
@@ -130,6 +158,92 @@ def readFromStream(
+                            )
+                            objs.append((cur_obj, clon))
+                            assert prev_obj is not None
+                            prev_obj[NameObject(k)] = clon.indirect_ref
```
```suggestion
                            prev_obj[NameObject(k)] = clon.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+        if isinstance(page, NullObject):
+            return None
+        if isinstance(page, int):
+            _i = reader.pages[page].indirect_ref
+        # elif isinstance(page, PageObject):
+        #    _i = page.indirect_ref
+        elif isinstance(page, DictionaryObject) and page.get("/Type", "") == "/Page":
+            _i = page.indirect_ref
```
```suggestion
            _i = reader.pages[page].indirect_reference
        # elif isinstance(page, PageObject):
        #    _i = page.indirect_reference
        elif isinstance(page, DictionaryObject) and page.get("/Type", "") == "/Page":
            _i = page.indirect_reference
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                or "/Dest" in ano
+            ):
+                if "/Dest" not in ano:
+                    outlist.append(ano.clone(self).indirect_ref)
+                else:
+                    d = ano["/Dest"]
+                    if isinstance(d, str):
+                        # it is a named dest
+                        if str(d) in self.get_named_dest_root():
+                            outlist.append(ano.clone(self).indirect_ref)
+                    else:
+                        d = cast("ArrayObject", d)
+                        p = self._get_cloned_page(d[0], pages, reader)
+                        if p is not None:
+                            anc = ano.clone(self, ignore_fields=("/Dest",))
+                            anc[NameObject("/Dest")] = ArrayObject([p] + d[1:])
+                            outlist.append(anc.indirect_ref)
+            else:
+                d = cast("DictionaryObject", ano["/A"])["/D"]
+                if isinstance(d, str):
+                    # it is a named dest
+                    if str(d) in self.get_named_dest_root():
+                        outlist.append(ano.clone(self).indirect_ref)
```
```suggestion
                    outlist.append(ano.clone(self).indirect_reference)
                else:
                    d = ano["/Dest"]
                    if isinstance(d, str):
                        # it is a named dest
                        if str(d) in self.get_named_dest_root():
                            outlist.append(ano.clone(self).indirect_reference)
                    else:
                        d = cast("ArrayObject", d)
                        p = self._get_cloned_page(d[0], pages, reader)
                        if p is not None:
                            anc = ano.clone(self, ignore_fields=("/Dest",))
                            anc[NameObject("/Dest")] = ArrayObject([p] + d[1:])
                            outlist.append(anc.indirect_reference)
            else:
                d = cast("DictionaryObject", ano["/A"])["/D"]
                if isinstance(d, str):
                    # it is a named dest
                    if str(d) in self.get_named_dest_root():
                        outlist.append(ano.clone(self).indirect_reference)
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -2084,6 +2197,533 @@ def add_annotation(self, page_number: int, annotation: Dict[str, Any]) -> None:
+                        cast("DictionaryObject", anc["/A"])[
+                            NameObject("/D")
+                        ] = ArrayObject([p] + d[1:])
+                        outlist.append(anc.indirect_ref)
```
```suggestion
                        outlist.append(anc.indirect_reference)
```

MartinThoma commented:
I'm starting to hate the `indirect_ref` renaming (without an IDE at least) :sweat_smile:

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1479,14 +1563,15 @@ def add_named_destination_object(
-        page_destination_ref = self._add_object(page_destination)
+        page_destination_ref = self._add_object(page_destination.dest_array)
+        self.add_named_destination_array(
+            cast("TextStringObject", dest["/Title"]), page_destination_ref
```
```suggestion
            cast("TextStringObject", page_destination["/Title"]), page_destination_ref
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, str):  # it means that we are on the old params
+            return self.add_outline_item(
+                title, pagenum, parent, before, color, bold, italic, fit,
```
```suggestion
                title, pagenum, parent, None, before, color, bold, italic, fit,
```

MartinThoma commented:
Only now I have seen your solution to make the argument change backwards compatible (ignoring potential mypy issues of users) :man_facepalming:  I would have merged it a lot earlier if I had noticed that :see_no_evil:

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
         :param Fit fit: The fit of the destination page.
         """
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, str):  # it means that we are on the old params
```
```suggestion
        if isinstance(italic, Fit):  # it means that we are on the old params
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, Fit):  # it means that we are on the old params
+            return self.add_outline_item(
+                title, pagenum, parent, None, before, color, bold, italic, fit,
```
```suggestion
                title, page_number, parent, None, before, color, bold, italic, fit, pagenum
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
-            warnings.warn(
-                message=(
-                    f"{old_term} is deprecated as an argument. Use {new_term} instead"
+        if pagenum is None:
```
```suggestion
        if page_number is None:
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
+        if pagenum is None:
+            action_ref = None
+        else:
+            if isinstance(pagenum, IndirectObject):
+                page_ref = pagenum
+            elif isinstance(pagenum, PageObject):
+                page_ref = pagenum.indirect_reference
+            elif isinstance(pagenum, int):
```
```suggestion
            if isinstance(page_number, IndirectObject):
                page_ref = page_number
            elif isinstance(page_number, PageObject):
                page_ref = page_number.indirect_reference
            elif isinstance(page_number, int):
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
+                page_ref = pagenum.indirect_reference
+            elif isinstance(pagenum, int):
+                try:
+                    page_ref = self.pages[pagenum].indirect_reference
+                except IndexError:
+                    page_ref = NumberObject(pagenum)
+            if page_ref is None:
+                logger_warning(
+                    f"can not find reference of page {pagenum}",
```
```suggestion
                    page_ref = self.pages[page_number].indirect_reference
                except IndexError:
                    page_ref = NumberObject(page_number)
            if page_ref is None:
                logger_warning(
                    f"can not find reference of page {page_number}",
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, Fit):  # it means that we are on the old params
+            return self.add_outline_item(
+                title, page_number, parent, None, before, color, bold, italic, fit, pagenum
```
```suggestion
                title, page_number, parent, None, before, color, bold, italic, fit
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1362,42 +1420,52 @@ def add_outline_item(
         :param Fit fit: The fit of the destination page.
         """
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, Fit):  # it means that we are on the old params
+            return self.add_outline_item(
+                title, page_number, parent, None, before, color, bold, italic, fit
+            )
```
```suggestion
        if isinstance(italic, Fit):  # it means that we are on the old params
            if fit is not None and page_number is None:
                page_number = fit
            return self.add_outline_item(
                title, page_number, parent, None, before, color, bold, italic
            )
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -32,27 +32,31 @@
     cast,
 )
 
+from ._encryption import Encryption
```
```suggestion
from ._encryption import Encryption
from .generic import PAGE_FIT
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -32,27 +32,32 @@
 )
 
+from ._encryption import Encryption
+from .generic import PAGE_FIT
```
```suggestion
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1364,42 +1420,54 @@ def add_outline_item(
+        page_ref: Union[None, NullObject, IndirectObject, NumberObject]
+        if isinstance(italic, Fit):  # it means that we are on the old params
+            if fit is not None and page_number is None:
+                page_number = fit
```
```suggestion
                page_number = fit  # type: ignore
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1364,42 +1420,54 @@ def add_outline_item(
+            if fit is not None and page_number is None:
+                page_number = fit
+            return self.add_outline_item(
+                title, page_number, parent, None, before, color, bold, italic
```
```suggestion
                title, page_number, parent, None, before, color, bold, italic  # type: ignore
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1481,14 +1565,15 @@ def add_named_destination_object(
             raise ValueError("page_destination may not be None")
 
-        page_destination_ref = self._add_object(page_destination)
+        page_destination_ref = self._add_object(page_destination.dest_array)
```
```suggestion
        page_destination_ref = self._add_object(page_destination.dest_array)  # type: ignore
```

MartinThoma commented on PyPDF2/_writer.py:
```diff
@@ -1481,14 +1565,15 @@ def add_named_destination_object(
-        page_destination_ref = self._add_object(page_destination)
+        page_destination_ref = self._add_object(page_destination.dest_array)
+        self.add_named_destination_array(
+            cast("TextStringObject", page_destination["/Title"]), page_destination_ref
```
```suggestion
            cast("TextStringObject", page_destination["/Title"]), page_destination_ref  # type: ignore
```

MartinThoma commented:
Finally! I'll have another quick look at the code and then merge today :-)

MartinThoma commented:
@pubpub-zz Thank you so much for this moonshot extension :pray: :heart:

xilopaint commented:
@pubpub-zz thanks for all the effort you've put into this PR!
